core-tests/jvm/src/test/scala/zio/SemaphoreSpec.scala

```diff
@@ -3,109 +3,97 @@
-      permits <- (s.release *> clock.sleep(10.milliseconds) *> s.available)
-    } yield permits must_=== 1L
-
-    unsafeRun(test)
-  }
-
+object SemaphoreSpecData {
   def offsettingReleasesAcquires(
     acquires: (Semaphore, Vector[Long]) => UIO[Unit],
     releases: (Semaphore, Vector[Long]) => UIO[Unit]
   ) = {
     val permits = Vector(1L, 0L, 20L, 4L, 0L, 5L, 2L, 1L, 1L, 3L)
 
-    unsafeRun(for {
+    for {
       semaphore     <- Semaphore.make(0L)
       acquiresFiber <- acquires(semaphore, permits).fork
       releasesFiber <- releases(semaphore, permits).fork
       _             <- acquiresFiber.join
       _             <- releasesFiber.join
       count         <- semaphore.available
-    } yield count must_=== 0)
+    } yield assert(count, equalTo(0L))
   }
-
 }
+
+object SemaphoreSpec
+    extends ZIOBaseSpec(
+      suite("SemaphoreSpec")(
+        suite("Make a Semaphore and verify that")(
+          testM("`acquire` permits sequentially") {
+            val n = 20L
+            for {
+              semaphore <- Semaphore.make(n)
+              available <- IO.foreach((0L until n).toList)(_ => semaphore.acquire) *> semaphore.available
+            } yield assert(available, equalTo(0L))
+          },
+          testM("`acquire` permits in parallel") {
+            val n = 20L
+            for {
+              semaphore <- Semaphore.make(n)
+              available <- IO.foreachPar((0L until n).toList)(_ => semaphore.acquire) *> semaphore.available
+            } yield assert(available, equalTo(0L))
+          },
+          testM("`acquireN`s can be parallel with `releaseN`s") {
+            offsettingReleasesAcquires(
+              (s, permits) => IO.foreach(permits)(s.acquireN).unit,
+              (s, permits) => IO.foreach(permits.reverse)(s.releaseN).unit
+            )
+          },
+          testM("individual `acquireN`s can be parallel with individual `releaseN`s") {
+            offsettingReleasesAcquires(
+              (s, permits) => IO.foreachPar(permits)(amount => s.acquireN(amount)).unit,
+              (s, permits) => IO.foreachPar(permits.reverse)(amount => s.releaseN(amount)).unit
+            )
+          },
+          testM("semaphores and fibers play ball together") {
+            val n = 1L
+            for {
+              s <- Semaphore.make(n).tap(_.acquire)
+              _ <- s.release.fork
+              _ <- s.acquire
+            } yield assert((), equalTo(()))
+          },
+          /**
+           * Ported from @mpilquist work in Cats Effect (https://github.com/typelevel/cats-effect/pull/403)
+           */
+          testM("`acquire` doesn't leak permits upon cancellation") {
+            val n = 1L
+            (for {
+              s       <- Semaphore.make(n)
+              _       <- s.acquireN(2).timeout(1.milli).either
+              permits <- s.release *> clock.sleep(10.millis) *> s.available
+            } yield assert(permits, equalTo(2L))).provide(Clock.Live)
+          },
```

Comment:
I've been trying to make this work with `MockClock` but any attempt just timed out after 1m.
sample code:
```
      testM("`acquire` doesn't leak permits upon cancellation") {
        for {
          mockClock <- MockClock.makeMock(MockClock.DefaultData)
          s         <- Semaphore.make(1L)
          sf         <- s.acquireN(2).timeout(1.milli).either.fork
          _         <- mockClock.adjust(1.second)
          _ <- sf.join
          _         <- s.release
          permits   <- s.available
        } yield assert(permits, equalTo(2L))
      },
```